### PR TITLE
Enable context isolation and disable nodejs support. Fixes #2018

### LINKED ISSUE #2018: Electron security considerations
The [security recommendations from electron](https://github.com/electron/electron/blob/master/docs/tutorial/security.md) outline best practice for securing the main process from malicious attacks in the renderer process. As electron is a little way behind chromium releases, known security holes may be publicised long before theia users receive updates to their apps. This is especially relevant if theia ever loads remote content.

Have these guidelines been considered in the design of theia and is there interest in applying them?

For example, I would recommend adding `webPreferences` to the `BrowserWindow` options for these items:

- `nodeIntegration: false` to prevent malicious code from running node functionality from the renderer
- `preload` to control the exact functionality and APIs available to the renderer process from the main process. This could simply be restricted to the IPC interfaces

This may have an impact on how `electron-browser` modules access the main electron process, however and require some re-architecture. I'm happy to propose a PR if this is agreed as being needed.

### PR DESCRIPTION
**Pull request by tsmaeder**
#### What it does
All access to electron API is now done through an API exposed via a preload script. Access to the electron API (including electron-remote) and nodejs API is no longer possible.
Theia extensions can contribute to the preload script via a `theiaExtensions` module declaration in their package.json

Fixes #2018

Contributed on behalf or STMicroelectronics

#### How to test
Run the electron version of Theia and make sure nothing is broken. Particular areas of interest is anything to do with menus and windows, for example "zoom in/out", "maximize", etc.

#### CQ

- [x] [Electron IP Check](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/7508)
- [x] [FFmpeg IP Check](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/7507)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by tsmaeder** [comment]
I'm opening a draft PR because I'll be working on updating to a newer electron version, as well. I'd be interested in feedback and testing, though.

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts
Comment (paul-marechal) [inline packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:187]: How can a "DTO" contain functions?

File: packages/filesystem/src/electron-browser/preload.ts
Comment (paul-marechal) [inline packages/filesystem/src/electron-browser/preload.ts:19]: Use the shared `electron` module, we should not rely on implicit dependencies.

**Review by paul-marechal — COMMENTED** [review]
Quick first review pass. Note that the CI doesn't look happy.

**Comment by tsmaeder** [comment]
The license check shows npm/npmjs/@babel/plugin-transform-async-to-generator/7.20.7 to be "under review", but if you looka the issue, it's approved. What gives? https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/6230

**Comment by tsmaeder** [comment]
One of the linter issues is this: 
```packages/debug/src/browser/debug-variables-source-tree.ts(46,11): error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'SourceTree'.
```

But that's not even part of this PR. What gives?

**Comment by vince-fugnitto** [comment]
> The license check shows npm/npmjs/@babel/plugin-transform-async-to-generator/7.20.7 to be "under review", but if you looka the issue, it's approved. What gives? https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/6230

@tsmaeder I don't believe these are errors, only `electron` is:

- https://github.com/eclipse-theia/theia/actions/runs/4436118782/jobs/7784138808?pr=12299#step:5:348

The other entries were added to our **baseline** at the time but can likely be cleaned up as they are approved:

https://github.com/eclipse-theia/theia/blob/f8654ec6b300b4215d48327d2500cbfb081c3bce/dependency-check-baseline.json#L1-L14

**Comment by tsmaeder** [comment]
@vince-fugnitto thanks for chiming in. Is there any documentation on what a "baseline" is and how to resolve issues with this check?

**Comment by tsmaeder** [comment]
FYI: I got the `dash-licenses` check to pass by poking at clearlydefined to harvest data for the electron versions in question. Not that I know what I'm doing 🤷

**Comment by vince-fugnitto** [comment]
> Do you have additional info that says we still need a CQ?

It is required due to Electron distributing `ffmpeg` which we both file a CQ for and have `@theia/ffmpeg` which performs a replacement to only include non-proprietary codecs.

- https://github.com/eclipse-theia/theia/blob/master/NOTICE.md#electron

The foundation has requested in the past that we submit requests to be able to use Electron in the project.
cc @marcdumais-work

**Comment by tsmaeder** [comment]
@paul-marechal you mentioned you would prefer using dependency injection for some of the API. Could you sketch what would do differently?

**Comment by paul-marechal** [comment]
Instead of creating hardcoded APIs and exposing them all in the global scope, we'd do like any other entry point and load various Inversify container modules. Doing this in the preload context was easy, but then it gets tricky when trying to expose the things bound there to the browser window.

Ultimately the idea is to go from referring to preload services as `window.whateverGlobalVariable` to be able to use the `@inject(PreloadComponentApi)` pattern.

This means that downstream app makers will be able to override services/apis/etc just like everywhere else. With this PR in its current state it's not possible to change much.

**Comment by tsmaeder** [comment]
@paul-marechal that's sounds interesting, but it's really in scope for this PR, IMO. It sounds like a complete rewrite of the code. I would much rather focus on getting this one merged and to proceed from there.

**Comment by tsmaeder** [comment]
In German, there is a saying: "the better is the enemy of the good" ;-)

**Comment by paul-marechal** [comment]
In software development there is a saying: "Temporary solutions are permanent" ;)

**Comment by paul-marechal** [comment]
> It sounds like a complete rewrite of the code. I would much rather focus on getting this one merged and to proceed from there.

If we agree now that we want a "complete rewrite of the code" then let's hold on merging this PR and take care of said rewrite. I don't see a reason to rush this feature in. I am also currently working on said rewrite so you can continue your work on the other matters you mentioned. My plan is not to rewrite everything but write enough that we have a stable foundation and have you continue porting the old code to the new pattern, if that's ok with you.

I'm currently experimenting with ways to properly interface Inversify containers from one context to the other via the Electron APIs, it is a struggle but I'm not out of things to try yet. My goal is to make it as ergonomic as possible for the end users of this isolation feature.

> "the better is the enemy of the good"

I don't feel like this applies to what I'm asking here: To me this PR won't be "good" as long as we don't get rid of the hardcoded logic. That's my threshold for "good enough" and I won't keep asking for "better", unless of course I keep seeing things that are legitimate issues and need fixing.

**Comment by tsmaeder** [comment]
@paul-marechal if I understand correctly, what you object to is that what API is exposed by a "preload" module cannot be overriden by adopters through dependency injection. I agree that this would be a desirable property. But is it a necessary first step? I don't think so. The utter lack of feedback on the topic of electron usage from adopters makes me think it is mostly a non-issue. If you want to improve the architecture, why not work on a follow-up PR? I'm not objecting and more than willing to review it.

>  I don't see a reason to rush this feature

Well, this PR is a dependency for enabling secondary window support in electron, due to https://github.com/electron/electron/issues/36858. That is a well-received differentiator feature vs. VS Code. Furthermore, this PR brings us into better alignment with the security recommendations for electron (which was the original issue raised by @thegecko).  So there is real value in this PR and if we don't _have_ to delay it, I think we should not.

**Comment by msujew** [comment]
@tsmaeder @paul-marechal Note that this issue also prevents merging https://github.com/eclipse-theia/theia/pull/12192 due to a bug in the node.js version used in Electron. I'd also argue that we should merge this PR rather sooner than later.

**Comment by marcdumais-work** [comment]
I agree that the iterative approach is reasonable in this case, since that this PR adds sufficient value to be considered an improvement on its own, even if further refinements would also be valuable.

**Comment by kittaakos** [comment]
Thanks for the great work!

> agree that this would be a desirable property. But is it a necessary first step? I don't think so.

If a downstream project can copy-paste and modify the preload from Theia, it's OK until there is proper DI support. If this PR prevents a downstream project from bumping the Theia version, it's not OK.

**Comment by parliament718** [comment]
Thank you for this PR! Very important stuff. Would a merge kick off builds for new theia packages? It would be a dream to see this happen.

**Comment by paul-marechal** [comment]
> prevents a downstream project from bumping the Theia version

What do you mean?

**Comment by tsmaeder** [comment]
> If a downstream project can copy-paste and modify the preload from Theia

There is a mechanism to contribute do the preload script, but the granularity right now is at the package level: if you include a package, you get it's contributions loaded into the preload script. I don't really see this as a problem, though: worst case, an overriden contribution is not using some api exposed via the `window` variable.

File: packages/core/src/electron-main/electron-main-application-module.ts
Comment (kittaakos) [inline packages/core/src/electron-main/electron-main-application-module.ts:47]: Can Theia use this instead? It's easier to customize downstream.

```js
bind(TheiaMainApi).toSelf().inSingletonScope();
bind(ElectronMainApplicationContribution).toService(TheiaMainApi);
```

Same for the filesystem API binding.

File: packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts
Comment (kittaakos) [inline packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts:138]: Is it still needed? Theia moves to electron 22+.

File: packages/navigator/src/electron-browser/electron-navigator-menu-contribution.ts
Comment (kittaakos) [inline packages/navigator/src/electron-browser/electron-navigator-menu-contribution.ts:27]: Leftover?

File: packages/core/src/electron-main/electron-main-application.ts
Comment (kittaakos) [inline packages/core/src/electron-main/electron-main-application.ts:310]: Why is this false?

----

I have noticed that you deleted a few comments pointing to GH issues; I think you can delete the `// https://github.com/eclipse-theia/theia/issues/2018` comment 👇 too as your PR is going to resolve it 😊

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts
Comment (tsmaeder) [reply packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:187]: `Menu` is heavily overloaded and I didn't want to introduce another, similarly named entity. I think it kinda makes sense, since the intent of the object is to be transfered to the main process (with some massaging).

File: packages/core/src/electron-main/electron-main-application-module.ts
Comment (tsmaeder) [reply packages/core/src/electron-main/electron-main-application-module.ts:47]: Yeah, extricating contributions is a pain.

File: packages/core/src/electron-main/electron-main-application.ts
Comment (tsmaeder) [reply packages/core/src/electron-main/electron-main-application.ts:310]: sandbox is `false`, because it's not part of either the original issue nor the requirements for detachable windows. Before Electron 20, sandboxing was disabled by default. I have not investigated what that step would require, nor extensively tested with the sandbox enabled. One step at a time.

**Comment by tsmaeder** [comment]
> Thanks for keeping me in the loop 🙏
> 
> For example, downstream wants to support electron menu icons. Currently, it's possible by subclassing the `electron-main-menu-factory.ts`.

I would argue that it's possible because electron-remote offers the whole electron `Menu` API instead of the reduced set that the `TheiaCoreAPI` offers. The right fix, IMO is to expose and `icon` field in the `MenuDto` class. 
> 
> After this PR, the `electron-main-menu-factory.ts` still needs to be subclassed, and the electron main part. All seems reasonable:
> 
> https://github.com/eclipse-theia/theia/blob/bcb49abae88d0652f01cebd2b245322b9011ffa0/packages/core/src/electron-main/electron-api-main.ts#L217-L226
> 
> Can the browser preload be customized here without patching the code after `theia build`?

You can customize the browser preload by contributing a `preload` module in your package.json. It's in the `Migration.md` document. As main/renderer API's are not designed to be subclassed, in your particular use case, you'd probably just contribute your own script to the preload that has extended, menu-related API. Your subclass of `ElectronMainMenuFactory` would then use that extended API instead of the one contributed by the Theia `core` module.

**Comment by kittaakos** [comment]
> I would argue that it's possible because

You don't need to argue. It's working with Theia.


https://user-images.githubusercontent.com/1405703/230005432-fb6bf0b6-2994-4606-b41d-7614bde517f2.mp4



> The right fix, IMO is to expose and `icon` field in the `MenuDto` class.

This is precisely what I was asking. Can downstream projects do it without Theia changes?

**Comment by tsmaeder** [comment]
> This is precisely what I was asking. Can downstream projects do it without Theia changes?

Yes, but as stated above: if you use electron-remote API, you need to implement that in your extensions like so: 

> You can customize the browser preload by contributing a preload module in your package.json. It's in the Migration.md document. As main/renderer API's are not designed to be subclassed, in your particular use case, you'd probably just contribute your own script to the preload that has extended, menu-related API. Your subclass of ElectronMainMenuFactory would then use that extended API instead of the one contributed by the Theia core module.

Your code seems to have used a feature in electron-remote that is not present in the main/renderer API that this PR exposes. Using icons in menus sounds like a reasonably feature to build into the `TheiaCoreAPI` main/rendere API, but since I'm not familiar with your code, I'd need some info on what you're actually using.

I seem not to give you the info you need to evaluate this PR, but I don't know how to make the idea clearer than this. Please advise?

**Comment by kittaakos** [comment]
> Please advise?

I propose to merge this PR as soon as all parties are OK. I cannot wait to see this change on the `master`. I am sure our security team here will also be pleased. Excellent work, thank you 👍 

Once the changes are on the `master`, I customize the default preload behavior from the `api-samples` extension. If it's not feasible, we will know what is required, and you could advise then, but no need to do anything now. What do you think?

**Comment by tsmaeder** [comment]
> Once the changes are on the master, I customize the default preload behavior from the api-samples extension. If it's not feasible, we will know what is required, and you could advise then, but no need to do anything now. What do you think?

Sounds good to me. This was always going to be a breaking change. Feedback from adopters on how to make it less disruptive is valuable and very welcome.

**Comment by parliament718** [comment]
Indeed. I've been checking back here every day, very eager to see this merged as Theia integration in our app is out of the question without `contextIsolation`. 

Will be happy to report back any issues with preload behavior as soon as I can install these new goodies. Thanks again for this PR

File: packages/core/src/electron-common/electron-api.ts
Comment (msujew) [inline packages/core/src/electron-common/electron-api.ts:15]:
```suggestion
// *****************************************************************************
```

File: packages/core/src/electron-common/electron-api.ts
Comment (msujew) [inline packages/core/src/electron-common/electron-api.ts:107]:
```suggestion
export const CHANNEL_GET_TITLE_STYLE_AT_STARTUP = 'GetTitleStyleAtStartup';
```

File: packages/filesystem/src/electron-browser/preload.ts
Comment (msujew) [inline packages/filesystem/src/electron-browser/preload.ts:24]:
```suggestion
```

File: packages/filesystem/src/electron-main/electron-main-module.ts
Comment (msujew) [inline packages/filesystem/src/electron-main/electron-main-module.ts:15]:
```suggestion
// *****************************************************************************
```

File: packages/navigator/src/electron-browser/electron-navigator-menu-contribution.ts
Comment (msujew) [inline packages/navigator/src/electron-browser/electron-navigator-menu-contribution.ts:27]:
```suggestion
import '@theia/core/lib/electron-common/electron-api';
```

**Review by msujew — COMMENTED** [review]
The changes look fairly good and I can confirm that most features continue to work as expected on Windows 11.

Also, I was able to confirm that this change indeed addresses the issue brought up in https://github.com/eclipse-theia/theia/pull/12192. I was successfully able to run the CSS language features extension with the updated version of Node/Electron in German.

However, I've noticed that one of the messages **doesn't work as expected**. When closing the application with a dirty file in the workspace, the application just closes without any confirmation message. This is a regression from `master`.

**Comment by tsmaeder** [comment]
> When closing the application with a dirty file in the workspace, the application just closes without any confirmation message. This is a regression from master.

I'll have a look, @msujew

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts
Comment (paul-marechal) [reply packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:187]: I had time to get familiar with the code as well as Electron's APIs, and Electron allows functions in objects passed between the browser and preload contexts.

I don't have a better naming for `MenuDto` either so I'm fine with it.

**Comment by paul-marechal** [comment]
Can't unmaximize a window if it was maximized before starting the app:

- Have the last window maximized before closing it.
- Restart the app.
- If the window correctly re-opens maximized, then the title bar control doesn't let you unmaximize.

Note that this issue is fixed in the alternative PR.

**Review by msujew — COMMENTED** [review]
I'm happy with the changes, and everything seems to work as expected now (on Windows) 👍 

I would like the approval of someone who can test the changes on other operating systems, before I can approve a merge of this PR though.

**Comment by kittaakos** [comment]
> who can test the changes on other operating systems,

> In particular Mac testing

I cannot approve/reject it, but I will give it a try on macOS later today. Thanks for your work, Thomas

**Comment by msujew** [comment]
@kittaakos I would be willing to approve this PR if you provide your approval.

**Review by msujew — APPROVED** [review]
Thanks @kittaakos! I guess the issue is indeed unrelated to the changes in this PR.

Alright, then I'll approve this. We will have time during the next few weeks to see whether we have introduced any other accidental regressions 👍

**Review by vince-fugnitto — DISMISSED** [review]
Just so we do not accidentally merge, we are still waiting for approval from the foundation regarding the IP Checks:
- [Electron IP Check](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/7508)
- [FFmpeg IP Check](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/7507)

Perhaps as we wait it's worthwhile to review https://github.com/eclipse-theia/theia/pull/12380 which improves upon the pull-request, uses inversify and fixes some [known issues](https://github.com/eclipse-theia/theia/pull/12380#issuecomment-1499422118)?

**Comment by tsmaeder** [comment]
The IP checks both have been approved.

**Comment by tsmaeder** [comment]
@vince-fugnitto I rebased on master. If you could remove your block, we're good to go.

**Review by vince-fugnitto — COMMENTED** [review]
I've noticed some odd behaviors to look out for, most likely as a follow-up pull-request:

image

The theia menu-icon is not rendered, the main-area placeholder icon is visibly distorted when resizing the application.

**Review by vince-fugnitto — COMMENTED** [review]
@tsmaeder can you confirm if you have this bad bug?

1. start the application
2. confirm that the "native" main-menu works as intended
3. maximise the window
4. the "native" main-menu stops working properly

It seems that the app is in some weird state where it does not have window focus resulting in menu items not working properly.

**Comment by tsmaeder** [comment]
> @tsmaeder can you confirm if you have this bad bug?

No, this works for me.

**Comment by vince-fugnitto** [comment]
> No, this works for me. Windows?

On Linux for me:

https://user-images.githubusercontent.com/40359487/233078130-8f4ca5ee-72ff-474d-8eb4-5a3532c9e5fd.mp4

Once I switch apps and come back to Theia the menu starts to work again, it's almost like maximizing/minimzing windows makes the window lose some focus.

**Comment by tsmaeder** [comment]
"Works for me" (TM)

https://user-images.githubusercontent.com/13163770/233084511-286deeda-036f-45f8-84f7-5d04fb141111.mp4

**Comment by tsmaeder** [comment]
@vince-fugnitto @marcdumais-work I guess you counter-checked that it doesn't matter on master?

**Comment by marcdumais-work** [comment]
> I guess you counter-checked that it doesn't matter on master?

Good suggestion - I will try now

**Comment by tsmaeder** [comment]
@vince-fugnitto your video is cut off at the top: you're really talking about "maximize", not "full-screen", right?

**Comment by tsmaeder** [comment]
Still works for me 🤷

**Comment by marcdumais-work** [comment]
> Still works for me shrug

Both Vince and myself use a Ubuntu 20.04 based OS. It could be the issue is specific to this. In any case it's more an annoyance than a serious bug IMHO.

**Comment by tsmaeder** [comment]
I guess no error messages in any logs?

**Comment by vince-fugnitto** [comment]
For reference I believe we are encountering https://github.com/electron/electron/issues/35724, a backport was made last week.

**Pull request merged**
